src/message/body.rs

```diff
@@ -0,0 +1,333 @@
+use std::io::{self, Write};
+use std::str;
+
+use crate::message::header::ContentTransferEncoding;
+
+#[derive(Debug, Clone)]
+pub struct Body(BodyInner);
+
+#[derive(Debug, Clone)]
+enum BodyInner {
+    Binary(Vec<u8>),
+    String(String),
+}
+
+impl Body {
+    #[inline]
+    pub fn len(&self) -> usize {
+        match &self.0 {
+            BodyInner::Binary(b) => b.len(),
+            BodyInner::String(s) => s.len(),
+        }
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        match &self.0 {
+            BodyInner::Binary(b) => b.is_empty(),
+            BodyInner::String(s) => s.is_empty(),
+        }
+    }
+
+    #[inline]
+    pub fn as_str(&self) -> Option<&str> {
+        match &self.0 {
+            BodyInner::Binary(_) => None,
+            BodyInner::String(s) => Some(s.as_ref()),
+        }
+    }
+
+    pub fn encoding(&self) -> ContentTransferEncoding {
+        match &self.0 {
+            BodyInner::String(s) if can_be_7bit(s) => ContentTransferEncoding::SevenBit,
+            // TODO: consider when base64 would be a better option because of output size
+            BodyInner::String(_) => ContentTransferEncoding::QuotedPrintable,
+            BodyInner::Binary(_) => ContentTransferEncoding::Base64,
+        }
+    }
+
+    pub fn encode(self, encoding: ContentTransferEncoding) -> Body {
+        match encoding {
+            ContentTransferEncoding::SevenBit => {
+                debug_assert!(
```

Comment:
Thinking more about this, panicking doesn't sound like a good thing to do. The encoding might be wrong for any number of reasons. We should probably have this return a `Result`, letting the user deal with it.